adapters/projects/robinos/prediction.go

```diff
@@ -0,0 +1,117 @@
+
+const (
+	RobinosAddress                = "0x9EF9c57754eD079D750016b802DcCD45d0AB66f8"
+	LogRewardDistributedSignature = "RewardDistributed(string,uint256[],address[])"
+)
+
+var SelectedMultiplierEvents = []string{"Prediction - $TAIKO price on Oct 20", "Serie A 24/25 - Juventus v. Lazio", "EPL 24/25 - Liverpool v. Chelsea", "EPL 24/25 - Man Utd v. Brentford"}
+
+type PredictionIndexer struct {
+	client    *ethclient.Client
+	addresses []common.Address
+}
+
+func NewPredictionIndexer(client *ethclient.Client, addresses []common.Address) *PredictionIndexer {
+	return &PredictionIndexer{
+		client:    client,
+		addresses: addresses,
+	}
+}
+
+var _ adapters.LogIndexer[adapters.Prediction] = &PredictionIndexer{}
+
+func (indexer *PredictionIndexer) Addresses() []common.Address {
+	return indexer.addresses
+}
+
+func (indexer *PredictionIndexer) Index(ctx context.Context, logs ...types.Log) ([]adapters.Prediction, error) {
+	var predictions []adapters.Prediction
+	for _, l := range logs {
+		if !indexer.isRewardDistributedEvent(l) {
+			continue
+		}
+
+		// Unpack the reward distributed event
+		var event struct {
+			EventCode    string
+			UserRewards  []*big.Int
+			WinningUsers []common.Address
+		}
+		if err := processRewardDistributedLog(l, &event); err != nil {
+			return nil, err
+		}
+
+		if !indexer.isSelectedMultiplierEvent(event.EventCode) {
+			continue
+		}
+
+		// Fetch block details
+		block, err := indexer.client.BlockByNumber(ctx, big.NewInt(int64(l.BlockNumber)))
+		if err != nil {
+			return nil, err
+		}
+
+		// Process each user and their corresponding reward
+		for i, user := range event.WinningUsers {
+			reward := event.UserRewards[i]
+
+			// Create a Prediction struct
+			prediction := adapters.Prediction{
+				User:          user,
+				TokenAmount:   reward,
+				TokenDecimals: adapters.TaikoTokenDecimals,
+				Token:         common.HexToAddress(adapters.TaikoTokenAddress),
+				EventCode:     event.EventCode,
+				BlockTime:     block.Time(),
+				BlockNumber:   block.Number().Uint64(),
+				TxHash:        l.TxHash,
+			}
+
+			predictions = append(predictions, prediction)
+		}
+	}
+
+	return predictions, nil
+}
+
+// Helper function to unpack the RewardDistributed event from the log
+func processRewardDistributedLog(vLog types.Log, event *struct {
+	EventCode    string
+	UserRewards  []*big.Int
+	WinningUsers []common.Address
+}) error {
+	contractABI, err := abi.JSON(strings.NewReader(robinos.RobinosABI))
+	if err != nil {
+		return err
+	}
+	return contractABI.UnpackIntoInterface(event, "RewardDistributed", vLog.Data)
+}
+
+// Check if the log is a RewardDistributed event
+func (indexer *PredictionIndexer) isRewardDistributedEvent(l types.Log) bool {
+	return len(l.Topics) > 0 && l.Topics[0].Hex() == crypto.Keccak256Hash([]byte(LogRewardDistributedSignature)).Hex()
+}
+
+func (indexer *PredictionIndexer) isSelectedMultiplierEvent(eventCode string) bool {
+	for _, selectedEvent := range SelectedMultiplierEvents {
```

Comment:
with the events in the struct we can update this to something like
`
for _, selectedEvent := range indexer.events {
`